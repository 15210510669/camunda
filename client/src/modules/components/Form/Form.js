/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a commercial license.
 * You may not use this file except in compliance with the commercial license.
 */

import React from 'react';

import classnames from 'classnames';

import './Form.scss';

export default function Form({compact, title, description, horizontal, ...props}) {
  return (
    <form {...props} className={classnames('Form', {compact, horizontal}, props.className)}>
      {title && <h3 className="formTitle">{title}</h3>}
      {description && <p className="formDescription">{description}</p>}
      {props.children}
    </form>
  );
}

Form.Group = function FormGroup({noSpacing, ...props}) {
  return (
    <div {...props} className={classnames('FormGroup', {noSpacing}, props.className)}>
      {props.children}
    </div>
  );
};

Form.InputGroup = function InputGroup(props) {
  return (
    <div {...props} className={classnames('InputGroup', props.className)}>
      {props.children}
    </div>
  );
};
