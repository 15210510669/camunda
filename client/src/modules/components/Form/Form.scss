.Form {
  .InputGroup {
    display: flex;
    align-items: flex-end;
    > *:not(:first-child) {
      margin-left: 10px;
    }

    > *:not(.Select) {
      flex: 1;
      min-width: 0px;
    }
  }

  input[type='checkbox'],
  input[type='radio'],
  .Switch {
    margin-right: 10px;

    + label {
      display: inline-block;
    }
  }

  > fieldset:first-child legend {
    padding: 0;
  }

  fieldset {
    border: none;
    margin: 0;
    padding: 0;

    legend {
      padding: 10px 0;
      font-weight: bold;
      margin-top: 0;
    }

    > * {
      width: 100%;
      margin-top: 10px;
    }
  }

  .FormGroup > label,
  fieldset > label,
  fieldset legend > label,
  .checkLabel {
    display: flex;
    align-items: baseline;
  }

  &:not(.compact) {
    margin-bottom: 20px;

    .formTitle {
      margin-top: 10px;
      font-weight: bold;
    }

    .formDescription {
      margin-top: 10px;
      font-size: 0.875em;
    }

    .ButtonGroup {
      padding-bottom: 10px;
    }

    .Message {
      margin-bottom: 10px;
    }

    .FormGroup {
      &:not(.noSpacing) {
        padding: 17px 0;
      }

      &:last-child {
        padding-bottom: 0;
      }

      .Labeled label .label + * {
        width: 100%;
      }

      input[type='checkbox'],
      input[type='radio'],
      .Switch {
        + label {
          padding: 10px 0px;
        }
      }
    }

    &.horizontal {
      width: 90%;

      .formTitle,
      .formDescription {
        margin-left: 30%;
      }

      .InfoMessage,
      .ErrorMessage {
        margin-left: 30%;
        margin-top: 3px;
      }

      input[type='checkbox'],
      input[type='radio'],
      .Switch {
        margin-left: 30%;
      }

      legend {
        margin-left: 30%;
        .Switch {
          margin-left: 0;
        }
      }

      .FormGroup {
        .Labeled label {
          display: flex;
          align-items: baseline;

          .label.before {
            width: 30%;
            padding-right: 10px;
            margin-bottom: 0;
            text-align: right;

            + * {
              width: 70%;
            }
          }
        }

        > *:not(.Labeled):not(.LabeledInput):not(fieldset):not(label):first-child {
          width: 30%;
          display: inline-block;
          padding-right: 10px;
          margin-bottom: 0;
          text-align: right;

          + * {
            display: inline-flex;
            width: 70%;
          }
        }

        > span:first-child {
          font-size: 0.875em;
        }

        > label,
        > .LabeledInput .label.after {
          display: block;
          padding: 10px 0px;
        }
      }

      .Message {
        width: 110%;
      }
    }
  }

  &.compact {
    font-size: 0.875em;

    .formTitle {
      font-weight: bold;
    }

    .formDescription {
      margin-top: 5px;
      font-size: 0.875em;
    }

    > fieldset:first-child legend,
    > div:first-child > fieldset:first-child legend {
      border: none;
      padding: 0;
    }

    fieldset {
      padding: 0 0 20px 0;

      legend {
        border-top: 1px solid #d3d3d3;
        padding: 14px 0 0 0;
      }
    }

    .FormGroup {
      &:not(.noSpacing) {
        padding: 5px 0 20px 0;
        &:not(:first-child) {
          border-top: 1px solid #d3d3d3;
          padding-top: 14px;
        }
      }

      &.noSpacing > label:only-child {
        padding: 5px 0 14px 0;
      }

      > label,
      > .LabeledInput {
        &:not(:first-child) {
          margin-top: 10px;
        }
      }
    }

    .Labeled {
      .label {
        margin-bottom: 5px;
        + * {
          width: 100%;
        }
      }
    }
  }
}
