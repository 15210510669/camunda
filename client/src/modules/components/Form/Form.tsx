/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {ComponentPropsWithoutRef, FormEvent} from 'react';

import classnames from 'classnames';

import './Form.scss';

interface FormProps extends Omit<ComponentPropsWithoutRef<'form'>, 'title'> {
  title?: string | JSX.Element[];
  description?: string | JSX.Element[];
  compact?: boolean;
  horizontal?: boolean;
  onSubmit?: (evt: FormEvent<HTMLFormElement>) => void;
}

export default function Form({
  compact,
  title,
  description,
  horizontal,
  onSubmit,
  ...props
}: FormProps) {
  return (
    <form
      {...props}
      onSubmit={(evt) => {
        evt.preventDefault();
        onSubmit?.(evt);
      }}
      className={classnames('Form', {compact, horizontal}, props.className)}
    >
      {title && <h3 className="formTitle">{title}</h3>}
      {description && <p className="formDescription">{description}</p>}
      {props.children}
    </form>
  );
}

interface FormGroupProps extends ComponentPropsWithoutRef<'div'> {
  noSpacing?: boolean;
}

Form.Group = function FormGroup({noSpacing, ...props}: FormGroupProps) {
  return (
    <div {...props} className={classnames('FormGroup', {noSpacing}, props.className)}>
      {props.children}
    </div>
  );
};

Form.InputGroup = function InputGroup(props: ComponentPropsWithoutRef<'div'>) {
  return (
    <div {...props} className={classnames('InputGroup', props.className)}>
      {props.children}
    </div>
  );
};
