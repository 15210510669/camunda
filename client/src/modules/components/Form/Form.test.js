/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import React from 'react';
import {shallow} from 'enzyme';

import Form from './Form';

it('should render without crashing', () => {
  shallow(<Form />);
});

it('should render a .Form className by default', () => {
  const node = shallow(<Form />);

  expect(node).toMatchSelector('.Form');
});

it('should merge and render additional classNames as provided as a property', () => {
  const node = shallow(<Form className="foo" />);

  expect(node).toMatchSelector('.Form.foo');
});

it('should render child elements and their props', () => {
  const node = shallow(
    <Form horizontal>
      <Form.Group>
        <label>
          <Form.InputGroup>
            <select id="input1">
              <option value="test">test</option>
            </select>
            <input id="input2" type="text" />
          </Form.InputGroup>
        </label>
      </Form.Group>
    </Form>
  );

  expect(node).toMatchSnapshot();
});

it('should prevent the default action on form submit events', () => {
  const spy = jest.fn();
  const evt = {preventDefault: jest.fn()};

  const node = shallow(<Form onSubmit={spy} />);
  node.simulate('submit', evt);

  expect(evt.preventDefault).toHaveBeenCalled();
  expect(spy).toHaveBeenCalledWith(evt);
});
