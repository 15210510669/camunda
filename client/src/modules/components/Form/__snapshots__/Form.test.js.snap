// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render child elements and their props 1`] = `
<form
  className="Form horizontal"
>
  <FormGroup>
    <label>
      <InputGroup>
        <select
          id="input1"
        >
          <option
            value="test"
          >
            test
          </option>
        </select>
        <input
          id="input2"
          type="text"
        />
      </InputGroup>
    </label>
  </FormGroup>
</form>
`;
